Description: Make the library install path configurable
 The upstream sources unconditionally install to
 CMAKE_INSTALL_PREFIX/lib/mariadb, whereas we want to install to
 /usr/lib/$(DEB_HOST_MULTIARCH).  This patch creates a new choice
 "DEB" for INSTALL_LAYOUT which creates a more appropriate installation
 layout for Debian packages.
Author: Daniel Schepler <schepler@debian.org>
Forwarded: https://lists.launchpad.net/maria-developers/msg07924.html

--- a/cmake/install.cmake
+++ b/cmake/install.cmake
@@ -12,6 +12,9 @@
 # - RPM
 # Built with default prefix=/usr
 #
+# - DEB
+# Install library to lib/architecture instead of lib/mariadb/
+#
 #
 # The following va+riables are used and can be overwritten
 # 
@@ -27,13 +30,13 @@ IF(NOT INSTALL_LAYOUT)
 ENDIF()
 
 SET(INSTALL_LAYOUT ${INSTALL_LAYOUT} CACHE
-  STRING "Installation layout. Currently supported options are DEFAULT (tar.gz and zip) and RPM")
+  STRING "Installation layout. Currently supported options are DEFAULT (tar.gz and zip), RPM and DEB")
 
 # On Windows we only provide zip and .msi. Latter one uses a different packager. 
 IF(UNIX)
   IF(INSTALL_LAYOUT MATCHES "RPM")
     SET(libmariadb_prefix "/usr")
-  ELSEIF(INSTALL_LAYOUT MATCHES "DEFAULT")
+  ELSEIF(INSTALL_LAYOUT MATCHES "DEFAULT|DEB")
     SET(libmariadb_prefix ${CMAKE_INSTALL_PREFIX})
   ENDIF()
 ENDIF()
@@ -43,7 +46,7 @@ IF(CMAKE_DEFAULT_PREFIX_INITIALIZED_BY_D
 ENDIF()
 
 # check if the specified installation layout is valid
-SET(VALID_INSTALL_LAYOUTS "DEFAULT" "RPM")
+SET(VALID_INSTALL_LAYOUTS "DEFAULT" "RPM" "DEB")
 LIST(FIND VALID_INSTALL_LAYOUTS "${INSTALL_LAYOUT}" layout_no)
 IF(layout_no EQUAL -1)
   MESSAGE(FATAL_ERROR "Invalid installation layout. Please specify one of the following layouts: ${VALID_INSTALL_LAYOUTS}")
@@ -60,6 +63,7 @@ ENDIF()
 # DEFAULT layout
 #
 SET(SUFFIX_INSTALL_DIR_DEFAULT "mariadb")
+SET(LIBSUFFIX_INSTALL_DIR_DEFAULT "mariadb")
 SET(BIN_INSTALL_DIR_DEFAULT "bin")
 SET(LIB_INSTALL_DIR_DEFAULT "lib")
 SET(INCLUDE_INSTALL_DIR_DEFAULT "include")
@@ -70,6 +74,7 @@ SET(PLUGIN_INSTALL_DIR_DEFAULT "lib/plug
 # RPM layout
 #
 SET(SUFFIX_INSTALL_DIR_RPM "mariadb")
+SET(LIBSUFFIX_INSTALL_DIR_RPM "mariadb")
 SET(BIN_INSTALL_DIR_RPM "bin")
 IF(CMAKE_SYSTEM_PROCESSOR MATCHES "x86_64")
   SET(LIB_INSTALL_DIR_RPM "lib64")
@@ -84,6 +89,18 @@ SET(DOCS_INSTALL_DIR_RPM "docs")
 SET(PLUGIN_INSTALL_DIR_RPM "lib/plugin")
 
 #
+# DEB layout
+#
+set(SUFFIX_INSTALL_DIR_DEB "mariadb")
+SET(LIBSUFFIX_INSTALL_DIR_DEB "${CMAKE_LIBRARY_ARCHITECTURE}")
+set(BIN_INSTALL_DIR_DEB "bin")
+set(LIB_INSTALL_DIR_DEB "lib")
+set(PLUGIN_INSTALL_DIRDIR_DEB "lib/${CMAKE_LIBRARY_ARCHITECTURE}/plugin")
+set(INCLUDE_INSTALL_DIR_DEB "include")
+set(DOCS_INSTALL_DIR_DEB "docs")
+set(PLUGIN_INSTALL_DIR_DEB "lib/${CMAKE_LIBRARY_ARCHITECTURE}/mariadb/plugin")
+
+#
 # Overwrite defaults
 #
 IF(LIB_INSTALL_DIR)
@@ -114,7 +131,7 @@ ELSE()
   SET(SUFFIX_INSTALL_DIR_${INSTALL_LAYOUT} ${SUFFIX_INSTALL_DIR})
 ENDIF()
 
-FOREACH(dir "BIN" "LIB" "INCLUDE" "DOCS" "PREFIX" "SUFFIX" "PLUGIN")
+FOREACH(dir "BIN" "LIB" "INCLUDE" "DOCS" "PREFIX" "SUFFIX" "LIBSUFFIX" "PLUGIN")
   SET(${dir}_INSTALL_DIR ${${dir}_INSTALL_DIR_${INSTALL_LAYOUT}})
   MARK_AS_ADVANCED(${dir}_INSTALL_DIR)
 ENDFOREACH()
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -316,7 +316,7 @@ ELSE()
 ENDIF()
 MESSAGE(STATUS "Installation layout: ${INSTALL_LAYOUT}")
 MESSAGE(STATUS "Include files will be installed in ${PREFIX_INSTALL_DIR}/${INCLUDE_INSTALL_DIR}/${SUFFIX_INSTALL_DIR}")
-MESSAGE(STATUS "Libraries will be installed in ${PREFIX_INSTALL_DIR}/${LIB_INSTALL_DIR}/${SUFFIX_INSTALL_DIR}")
+MESSAGE(STATUS "Libraries will be installed in ${PREFIX_INSTALL_DIR}/${LIB_INSTALL_DIR}/${LIBSUFFIX_INSTALL_DIR}")
 MESSAGE(STATUS "Binaries will be installed in ${PREFIX_INSTALL_DIR}/${BIN_INSTALL_DIR}")
 
 MESSAGE(STATUS "Required: ${CMAKE_REQUIRED_LIBRARIES}")
--- a/libmariadb/CMakeLists.txt
+++ b/libmariadb/CMakeLists.txt
@@ -420,7 +420,7 @@ INCLUDE(${CMAKE_SOURCE_DIR}/cmake/symlin
 # of the config program. To make sure these programs can
 # use mariadb client library we provide libmysql symlinks
 IF(NOT WIN32 AND WITH_MYSQLCOMPAT)
-  SET(INSTALL_PATH ${LIB_INSTALL_DIR}/${SUFFIX_INSTALL_DIR})
+  SET(INSTALL_PATH ${LIB_INSTALL_DIR}/${LIBSUFFIX_INSTALL_DIR})
   create_symlink(libmysqlclient${CMAKE_SHARED_LIBRARY_SUFFIX} libmariadb ${INSTALL_PATH})
   create_symlink(libmysqlclient_r${CMAKE_SHARED_LIBRARY_SUFFIX} libmariadb ${INSTALL_PATH})
   create_symlink(libmysqlclient${CMAKE_STATIC_LIBRARY_SUFFIX} mariadbclient ${INSTALL_PATH})
@@ -430,9 +430,9 @@ ENDIF()
 
 INSTALL(TARGETS
           libmariadb mariadbclient
-          RUNTIME DESTINATION "${LIB_INSTALL_DIR}/${SUFFIX_INSTALL_DIR}"
-          LIBRARY DESTINATION "${LIB_INSTALL_DIR}/${SUFFIX_INSTALL_DIR}"
-          ARCHIVE DESTINATION "${LIB_INSTALL_DIR}/${SUFFIX_INSTALL_DIR}")
+          RUNTIME DESTINATION "${LIB_INSTALL_DIR}/${LIBSUFFIX_INSTALL_DIR}"
+          LIBRARY DESTINATION "${LIB_INSTALL_DIR}/${LIBSUFFIX_INSTALL_DIR}"
+          ARCHIVE DESTINATION "${LIB_INSTALL_DIR}/${LIBSUFFIX_INSTALL_DIR}")
 
 INSTALL(DIRECTORY ${CMAKE_SOURCE_DIR}/include/
         DESTINATION ${INCLUDE_INSTALL_DIR}/${SUFFIX_INSTALL_DIR})
--- a/mariadb_config/mariadb_config.c.in
+++ b/mariadb_config/mariadb_config.c.in
@@ -4,7 +4,7 @@
 #include <stdio.h>
 
 #define INCLUDE "-I@PREFIX_INSTALL_DIR@/@INCLUDE_INSTALL_DIR@/@SUFFIX_INSTALL_DIR@ -I@PREFIX_INSTALL_DIR@/@INCLUDE_INSTALL_DIR@/@SUFFIX_INSTALL_DIR@/mysql"
-#define LIBS    "-L@PREFIX_INSTALL_DIR@/@LIB_INSTALL_DIR@/@SUFFIX_INSTALL_DIR@ -lmariadb" \
+#define LIBS    "-L@PREFIX_INSTALL_DIR@/@LIB_INSTALL_DIR@/@LIBSUFFIX_INSTALL_DIR@ -lmariadb" \
                 "@extra_dynamic_LDFLAGS@"
 #define CFLAGS  INCLUDE " @CMAKE_C_FLAGS@"
 #define VERSION "@MYSQL_CLIENT_VERSION@"
